### PR TITLE
Caching 20230213

### PR DESCRIPTION
**Pull request by carrolp**
Prevent parallel FetchEnv gets for the same kube resource, for optimal cache utilization.

### CODE REVIEW

File: lib/FetchEnvs.js [hunk 1]
```diff
@@ -160,23 +162,34 @@ module.exports = class FetchEnvs {
...
-      const krm = await this.kubeClass.getKubeResourceMeta(apiVersion, kind, 'update');
 
-      if (krm) {
-        try {
-          resource = await krm.get(name, namespace);
-          if( resource ) {
-            this.resourceCache.set( cacheKey, resource ); // Cache this resource
+    // Ensure same resource is not being retrieved multiple times simultaneously, for optimal cache utilization.
+    let limit;
+    if( !Object.prototype.hasOwnProperty.call( singleResourcePLimits, cacheKey ) ) singleResourcePLimits[cacheKey] = pLimit(1);
```
Comment (adamkingit) [inline lib/FetchEnvs.js:168]:
```suggestion
    if( !Object.prototype.hasOwnProperty.call( singleResourcePLimits, cacheKey ) )  {             
      singleResourcePLimits[cacheKey] = pLimit(1);
    }
```

File: lib/FetchEnvs.js [hunk 2]
```diff
@@ -228,42 +227,71 @@ module.exports = class FetchEnvs {
...
+        kubeError = error;
+      }
+    }
+    // Single-resource queries are cacheable.  If in the cache, use it.
+    else if( this.resourceCache.has( cacheKey ) ) {
+      response = this.resourceCache.get( cacheKey );
+    }
+    // Single-resource queries are cacheable.  If not in the cache, start an api call to populate the cache if needed, wait for it to finish, then use it from the cache.
+    else {
+      if( !singleResourceQueryCache[cacheKey] ) {
```
Comment (adamkingit) [inline lib/FetchEnvs.js:263]: Could we call #genericMapRef here to get the map, avoiding what seems like duplicate code?

File: lib/FetchEnvs.js [hunk 3]
```diff
@@ -29,14 +29,15 @@ const KIND_MAP = new Map([
 
 const LRU = require('lru-cache');
 const LruOptions = {
-  maxSize: 100000, // the max cache size
+  maxSize: parseInt(process.env.FETCHENVS_CACHE_SIZE) || 100000, // the max cache size
```
Comment (adamkingit) [inline lib/FetchEnvs.js:32]: remember to allow this to be set as in https://github.com/razee-io/MustacheTemplate/blob/master/kubernetes/MustacheTemplate/resource.yaml#L65-L70

File: lib/FetchEnvs.js (hunk 2 above)
Comment (carrolp) [reply lib/FetchEnvs.js]: `genericKeyRef can't directly call `genericMapRef` without additional refactoring, but combined the common function into `getSingleResource` function used in both places.

**Review by adamkingit — APPROVED** [review]

**Pull request merged**